### DISCUSSION TITLE
Extending the `FastifyRequest` interface conditionally

### DISCUSSION THREAD

**Discussion by are** [fastify/fastify#5620] [Q&A]
I was wondering whether there is a way of extending the type definition of the `req` object in a shorthand route definition based on the `config` (or any other property).

Example:
```ts
server.post("/posts", {
  schema: SomeSchema,
  config: { auth: "required" },
  async handler(req) {
    // req.user extends User
  }
});


server.get("/posts", {
  schema: SomeSchema,
  config: { auth: "optional" },
  async handler(req) {
    // req.user extends User | null
  }
});
```

**Comment by metcoder95** [comment]
I've not tested this, but I believe its "_doable_" if doing interface augmentation on the `FastifyRequest` based on a previously augmented `FastifyContextConfig`.

**Comment by MatyasZERO** [comment]
I found working solution:
```
import fastify, { ContextConfigDefault } from 'fastify';


type optional = "optional"
type required = "required"


interface User {
    name: string;
    age: number;
}


declare module 'fastify' {
  interface FastifyRequest<
    ContextConfig = ContextConfigDefault,
  > {
    user: ContextConfig["auth"] extends optional ? (User | null) : (ContextConfig["auth"] extends required ? User : never);
  }

  interface FastifyContextConfig {
    auth: optional | required;
  }
}
```

**Comment by wbizmo** [comment]
Yes, you can model this with Fastify’s `ContextConfig` generic.

The important part is that the route `config` type can flow into `FastifyRequest`, so `req.user` can be different depending on the route config.

Example:

```ts
import fastify, { ContextConfigDefault } from 'fastify'

type AuthMode = 'required' | 'optional'

interface User {
  id: string
  email: string
}

declare module 'fastify' {
  interface FastifyContextConfig {
    auth?: AuthMode
  }

  interface FastifyRequest<
    RouteGeneric,
    RawServer,
    RawRequest,
    SchemaCompiler,
    TypeProvider,
    ContextConfig = ContextConfigDefault
  > {
    user: ContextConfig extends { auth: 'required' }
      ? User
      : ContextConfig extends { auth: 'optional' }
        ? User | null
        : never
  }
}
```

Then your routes can express the difference through `config`:

```ts
server.post('/posts', {
  config: { auth: 'required' },
  async handler(req) {
    // req.user should be User here
    req.user.email
  }
})

server.get('/posts', {
  config: { auth: 'optional' },
  async handler(req) {
    // req.user should be User | null here
    if (req.user) {
      req.user.email
    }
  }
})
```

One small thing I would change from the earlier example is making `auth` optional on `FastifyContextConfig`.

That way routes without auth config do not have to define it.

Runtime-wise, you still need a hook/preHandler that actually sets `req.user`; this only solves the TypeScript side.

So the pattern is:

1. Add `auth` to `FastifyContextConfig`
2. Use the request `ContextConfig` generic to infer the user type
3. Add a hook/preHandler that populates `req.user` based on the route config

That should give you the conditional typing you're after.
